File: services/arcium-service/src/mpc/encryption.rs
use chacha20poly1305::{
    aead::{Aead, KeyInit, OsRng},
    ChaCha20Poly1305, Nonce,
};
use hkdf::Hkdf;
use rand::RngCore;
use sha2::Sha256;
use std::{error::Error, sync::OnceLock};

#[derive(Clone, Copy, Debug)]
enum EncryptionMode {
    Dev,
    Rescue,
}

fn resolve_mode() -> EncryptionMode {
    match std::env::var("ARCIUM_ENCRYPTION_BACKEND")
        .unwrap_or_else(|_| "dev".to_string())
        .to_lowercase()
        .as_str()
    {
        "rescue" => EncryptionMode::Rescue,
        _ => EncryptionMode::Dev,
    }
}

fn log_rescue_placeholder() {
    static RESCUE_LOG_ONCE: OnceLock<()> = OnceLock::new();
    RESCUE_LOG_ONCE.get_or_init(|| {
        log::warn!(
            "🪄 Rescue encryption backend requested. Falling back to development cipher until \
            native Rescue bindings are linked."
        );
    });
}

/// Development-mode encryption helper for MPC computations
///
/// ⚠️ WARNING: This is a DEVELOPMENT-MODE encryption implementation
/// Production deployment MUST use RescueCipher from Arcium SDK
///
/// This module provides:
/// - ChaCha20-Poly1305 AEAD encryption
/// - HKDF-based key derivation per user
/// - Compatible interface with future RescueCipher integration
///
/// Security model:
/// - Master key derived from environment secret
/// - Per-user keys derived via HKDF(master_key, user_pubkey)
/// - Random nonces for each encryption operation
/// - Authenticated encryption prevents tampering
///
/// Data format: [nonce (12 bytes)] + [ciphertext] + [auth tag (16 bytes)]
#[derive(Clone)]
pub struct EncryptionHelper {
    master_key: [u8; 32],
    mode: EncryptionMode,
}

impl EncryptionHelper {
    /// Create new encryption helper with master key
    ///
    /// The master key should be a 32-byte secret from environment config
    pub fn new_with_key(master_key: [u8; 32]) -> Self {
        let mode = resolve_mode();
        log::debug!("🔐 Encryption helper initialized ({:?} mode)", mode);
        Self { master_key, mode }
    }

    /// Create with default key (for testing only)
    pub fn new() -> Self {
        log::warn!("⚠️  Using default encryption key - NOT FOR PRODUCTION");
        Self {
            master_key: [42u8; 32], // Deterministic for testing
            mode: EncryptionMode::Dev,
        }
    }

    /// Derive a user-specific encryption key using HKDF
    ///
    /// Derivation: HKDF-SHA256(master_key, salt=user_pubkey, info="ninjapay-dev-v1")
    fn derive_user_key(&self, user_pubkey: &str) -> Result<[u8; 32], Box<dyn Error>> {
        let hkdf = Hkdf::<Sha256>::new(Some(user_pubkey.as_bytes()), &self.master_key);
        let mut derived_key = [0u8; 32];
        hkdf.expand(b"ninjapay-dev-v1", &mut derived_key)
            .map_err(|e| format!("Key derivation failed: {}", e))?;
        Ok(derived_key)
    }

    /// Encrypt a u64 value (for simulator)
    ///
    /// Converts u64 to 8-byte little-endian, then encrypts
    /// Returns: [nonce (12)] + [ciphertext (8)] + [tag (16)] = 36 bytes total
    pub fn encrypt_u64(&self, value: u64, user_pubkey: &str) -> Result<Vec<u8>, Box<dyn Error>> {
        let plaintext = value.to_le_bytes();
        self.encrypt_bytes(&plaintext, user_pubkey)
    }

    /// Decrypt to u64 value (for simulator)
    ///
    /// Decrypts and converts 8-byte little-endian to u64
    pub fn decrypt_to_u64(
        &self,
        encrypted: &[u8],
        user_pubkey: &str,
    ) -> Result<u64, Box<dyn Error>> {
        let plaintext = self.decrypt_bytes(encrypted, user_pubkey)?;
        if plaintext.len() != 8 {
            return Err(format!(
                "Invalid u64 decryption: expected 8 bytes, got {}",
                plaintext.len()
            )
            .into());
        }
        let bytes: [u8; 8] = plaintext
            .try_into()
            .map_err(|_| "Failed to convert to u64")?;
        Ok(u64::from_le_bytes(bytes))
    }

    /// Encrypt arbitrary bytes
    ///
    /// Format: [nonce (12 bytes)] + [ciphertext] + [tag (16 bytes)]
    pub fn encrypt_bytes(&self, data: &[u8], user_pubkey: &str) -> Result<Vec<u8>, Box<dyn Error>> {
        match self.mode {
            EncryptionMode::Dev => self.encrypt_bytes_dev(data, user_pubkey),
            EncryptionMode::Rescue => {
                log_rescue_placeholder();
                self.encrypt_bytes_dev(data, user_pubkey)
            }
        }
    }

    /// Decrypt arbitrary bytes
    ///
    /// Expects format: [nonce (12 bytes)] + [ciphertext] + [tag (16 bytes)]
    pub fn decrypt_bytes(
        &self,
        encrypted: &[u8],
        user_pubkey: &str,
    ) -> Result<Vec<u8>, Box<dyn Error>> {
        match self.mode {
            EncryptionMode::Dev => self.decrypt_bytes_dev(encrypted, user_pubkey),
            EncryptionMode::Rescue => {
                log_rescue_placeholder();
                self.decrypt_bytes_dev(encrypted, user_pubkey)
            }
        }
    }

    /// Validate encrypted input format
    ///
    /// Checks:
    /// 1. Not empty
    /// 2. Minimum length (nonce + tag = 28 bytes)
    /// 3. Proper structure
    pub fn validate_encrypted_input(&self, encrypted_data: &[u8]) -> Result<bool, Box<dyn Error>> {
        if encrypted_data.is_empty() {
            log::warn!("Validation failed: empty input");
            return Ok(false);
        }

        // Minimum: nonce (12) + minimal plaintext (0) + tag (16) = 28 bytes
        if encrypted_data.len() < 28 {
            log::warn!(
                "Validation failed: too short ({} bytes, need at least 28)",
                encrypted_data.len()
            );
            return Ok(false);
        }

        // For u64 encrypted values, expect exactly 36 bytes
        // nonce (12) + plaintext (8) + tag (16) = 36
        if encrypted_data.len() == 36 {
            log::debug!("✅ Valid u64 encrypted format");
            return Ok(true);
        }

        // Other lengths are valid as long as >= 28
        log::debug!("✅ Valid encrypted format ({} bytes)", encrypted_data.len());
        Ok(true)
    }

    /// Prepare encrypted inputs for batching
    ///
    /// Format: [count (4 bytes)] + [len1 (4)] + [data1] + [len2 (4)] + [data2] + ...
    pub fn prepare_batch_inputs(
        &self,
        encrypted_values: Vec<Vec<u8>>,
    ) -> Result<Vec<u8>, Box<dyn Error>> {
        // Validate all inputs first
        for (i, value) in encrypted_values.iter().enumerate() {
            if !self.validate_encrypted_input(value)? {
                return Err(format!("Invalid encrypted input at index {}", i).into());
            }
        }

        // Calculate total size
        let count = encrypted_values.len() as u32;
        let total_size: usize = 4 + // count
            encrypted_values.iter().map(|v| 4 + v.len()).sum::<usize>(); // len + data per item

        let mut result = Vec::with_capacity(total_size);

        // Write count
        result.extend_from_slice(&count.to_le_bytes());

        // Write each value with length prefix
        for value in encrypted_values {
            let len = value.len() as u32;
            result.extend_from_slice(&len.to_le_bytes());
            result.extend_from_slice(&value);
        }

        log::info!(
            "✅ Prepared batch of {} encrypted inputs ({} bytes)",
            count,
            result.len()
        );
        Ok(result)
    }

    fn encrypt_bytes_dev(&self, data: &[u8], user_pubkey: &str) -> Result<Vec<u8>, Box<dyn Error>> {
        let user_key = self.derive_user_key(user_pubkey)?;
        let cipher = ChaCha20Poly1305::new(&user_key.into());

        let mut nonce_bytes = [0u8; 12];
        OsRng.fill_bytes(&mut nonce_bytes);
        let nonce = Nonce::from_slice(&nonce_bytes);

        let ciphertext = cipher
            .encrypt(nonce, data)
            .map_err(|e| format!("Encryption failed: {}", e))?;

        let mut result = Vec::with_capacity(12 + ciphertext.len());
        result.extend_from_slice(&nonce_bytes);
        result.extend_from_slice(&ciphertext);

        log::debug!("✅ Encrypted {} bytes → {} bytes", data.len(), result.len());
        Ok(result)
    }

    fn decrypt_bytes_dev(
        &self,
        encrypted: &[u8],
        user_pubkey: &str,
    ) -> Result<Vec<u8>, Box<dyn Error>> {
        if encrypted.len() < 12 + 16 {
            return Err(format!(
                "Invalid encrypted data: too short (got {} bytes, need at least 28)",
                encrypted.len()
            )
            .into());
        }

        let (nonce_bytes, ciphertext) = encrypted.split_at(12);
        let nonce = Nonce::from_slice(nonce_bytes);

        let user_key = self.derive_user_key(user_pubkey)?;
        let cipher = ChaCha20Poly1305::new(&user_key.into());

        let plaintext = cipher
            .decrypt(nonce, ciphertext)
            .map_err(|e| format!("Decryption failed: {}", e))?;

        log::debug!(
            "✅ Decrypted {} bytes → {} bytes",
            encrypted.len(),
            plaintext.len()
        );
        Ok(plaintext)
    }

    /// Extract individual results from batch computation
    ///
    /// Expects format: [count (4 bytes)] + [len1 (4)] + [data1] + [len2 (4)] + [data2] + ...
    pub fn extract_batch_results(
        &self,
        batch_result: Vec<u8>,
        expected_count: usize,
    ) -> Result<Vec<Vec<u8>>, Box<dyn Error>> {
        if batch_result.len() < 4 {
            return Err("Batch result too short: missing count".into());
        }

        // Read count
        let count_bytes: [u8; 4] = batch_result[0..4]
            .try_into()
            .map_err(|_| "Failed to read count")?;
        let count = u32::from_le_bytes(count_bytes) as usize;

        if count != expected_count {
            return Err(format!(
                "Batch count mismatch: expected {}, got {}",
                expected_count, count
            )
            .into());
        }

        let mut results = Vec::with_capacity(count);
        let mut offset = 4;

        // Read each value
        for i in 0..count {
            if offset + 4 > batch_result.len() {
                return Err(format!("Batch result truncated at item {}", i).into());
            }

            // Read length
            let len_bytes: [u8; 4] = batch_result[offset..offset + 4]
                .try_into()
                .map_err(|_| format!("Failed to read length at item {}", i))?;
            let len = u32::from_le_bytes(len_bytes) as usize;
            offset += 4;

            // Read data
            if offset + len > batch_result.len() {
                return Err(format!("Batch result truncated: item {} data incomplete", i).into());
            }

            let data = batch_result[offset..offset + len].to_vec();
            results.push(data);
            offset += len;
        }

        log::info!("✅ Extracted {} results from batch", results.len());
        Ok(results)
    }

    /// Verify computation signature (placeholder for cluster mode)
    ///
    /// In development mode, this is a no-op
    /// In cluster mode, this verifies Arcium MPC cluster signatures
    pub fn verify_computation_signature(
        &self,
        _result: &[u8],
        _signature: &str,
        cluster_pubkey: &str,
    ) -> Result<bool, Box<dyn Error>> {
        log::debug!(
            "⚠️  Signature verification skipped (dev mode) for cluster: {}",
            cluster_pubkey
        );
        Ok(true)
    }
}

impl Default for EncryptionHelper {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_encrypt_decrypt_u64() {
        let helper = EncryptionHelper::new();
        let user = "test_user_pubkey";
        let value = 42u64;

        let encrypted = helper.encrypt_u64(value, user).unwrap();
        assert_eq!(encrypted.len(), 36); // nonce (12) + data (8) + tag (16)

        let decrypted = helper.decrypt_to_u64(&encrypted, user).unwrap();
        assert_eq!(decrypted, value);
    }

    #[test]
    fn test_encrypt_decrypt_bytes() {
        let helper = EncryptionHelper::new();
        let user = "test_user_pubkey";
        let data = b"Hello, NinjaPay!";

        let encrypted = helper.encrypt_bytes(data, user).unwrap();
        assert!(encrypted.len() > data.len()); // Should be longer (nonce + tag)

        let decrypted = helper.decrypt_bytes(&encrypted, user).unwrap();
        assert_eq!(decrypted, data);
    }

    #[test]
    fn test_different_users_different_ciphertexts() {
        let helper = EncryptionHelper::new();
        let value = 100u64;

        let encrypted1 = helper.encrypt_u64(value, "user1").unwrap();
        let encrypted2 = helper.encrypt_u64(value, "user2").unwrap();

        // Different users should produce different ciphertexts
        assert_ne!(encrypted1, encrypted2);

        // But decrypt to same value
        assert_eq!(helper.decrypt_to_u64(&encrypted1, "user1").unwrap(), value);
        assert_eq!(helper.decrypt_to_u64(&encrypted2, "user2").unwrap(), value);
    }

    #[test]
    fn test_wrong_user_cannot_decrypt() {
        let helper = EncryptionHelper::new();
        let encrypted = helper.encrypt_u64(42, "alice").unwrap();

        // Bob cannot decrypt Alice's data (will fail auth tag check)
        let result = helper.decrypt_to_u64(&encrypted, "bob");
        assert!(result.is_err());
    }

    #[test]
    fn test_validate_encrypted_input() {
        let helper = EncryptionHelper::new();

        // Empty input
        assert!(!helper.validate_encrypted_input(&[]).unwrap());

        // Too short
        assert!(!helper.validate_encrypted_input(&[1, 2, 3]).unwrap());

        // Valid encrypted u64
        let encrypted = helper.encrypt_u64(42, "test").unwrap();
        assert!(helper.validate_encrypted_input(&encrypted).unwrap());

        // Valid encrypted bytes
        let encrypted = helper.encrypt_bytes(b"test data", "test").unwrap();
        assert!(helper.validate_encrypted_input(&encrypted).unwrap());
    }

    #[test]
    fn test_tampering_detection() {
        let helper = EncryptionHelper::new();
        let mut encrypted = helper.encrypt_u64(42, "test").unwrap();

        // Tamper with ciphertext
        encrypted[20] ^= 0xFF;

        // Should fail decryption (auth tag verification)
        let result = helper.decrypt_to_u64(&encrypted, "test");
        assert!(result.is_err());
    }

    #[test]
    fn test_batch_prepare_extract() {
        let helper = EncryptionHelper::new();
        let user = "test";

        // Create batch of encrypted values
        let values = vec![
            helper.encrypt_u64(10, user).unwrap(),
            helper.encrypt_u64(20, user).unwrap(),
            helper.encrypt_u64(30, user).unwrap(),
        ];

        // Prepare batch
        let batched = helper.prepare_batch_inputs(values.clone()).unwrap();

        // Extract batch
        let extracted = helper.extract_batch_results(batched, 3).unwrap();

        // Should match original
        assert_eq!(extracted.len(), values.len());
        for (i, (original, extracted)) in values.iter().zip(extracted.iter()).enumerate() {
            assert_eq!(original, extracted, "Mismatch at index {}", i);
        }
    }

    #[test]
    fn test_batch_count_mismatch() {
        let helper = EncryptionHelper::new();
        let values = vec![
            helper.encrypt_u64(10, "test").unwrap(),
            helper.encrypt_u64(20, "test").unwrap(),
        ];

        let batched = helper.prepare_batch_inputs(values).unwrap();

        // Request wrong count
        let result = helper.extract_batch_results(batched, 3);
        assert!(result.is_err());
    }

    #[test]
    fn test_key_derivation_deterministic() {
        let helper = EncryptionHelper::new();

        // Same user should derive same key
        let key1 = helper.derive_user_key("alice").unwrap();
        let key2 = helper.derive_user_key("alice").unwrap();
        assert_eq!(key1, key2);

        // Different users should derive different keys
        let key_bob = helper.derive_user_key("bob").unwrap();
        assert_ne!(key1, key_bob);
    }

    #[test]
    fn test_nonce_randomness() {
        let helper = EncryptionHelper::new();

        // Same value encrypted twice should produce different ciphertexts (random nonces)
        let enc1 = helper.encrypt_u64(42, "test").unwrap();
        let enc2 = helper.encrypt_u64(42, "test").unwrap();
        assert_ne!(enc1, enc2);

        // But both should decrypt to same value
        assert_eq!(helper.decrypt_to_u64(&enc1, "test").unwrap(), 42);
        assert_eq!(helper.decrypt_to_u64(&enc2, "test").unwrap(), 42);
    }

    #[test]
    fn test_large_batch() {
        let helper = EncryptionHelper::new();
        let user = "test";

        // Create large batch (simulate payroll)
        let mut values = Vec::new();
        for i in 0..100 {
            values.push(helper.encrypt_u64(1000 + i, user).unwrap());
        }

        let batched = helper.prepare_batch_inputs(values.clone()).unwrap();
        let extracted = helper.extract_batch_results(batched, 100).unwrap();

        assert_eq!(extracted.len(), 100);
        for (i, (original, extracted)) in values.iter().zip(extracted.iter()).enumerate() {
            assert_eq!(original, extracted, "Mismatch at index {}", i);
        }
    }
}

File: packages/solana-utils/src/encryption.ts
import { chacha20poly1305 } from '@noble/ciphers/chacha';
import { randomBytes } from '@noble/ciphers/webcrypto';
import { hkdf } from '@noble/hashes/hkdf';
import { sha256 } from '@noble/hashes/sha256';

/**
 * Development-mode encryption for NinjaPay MPC computations
 *
 * ⚠️ WARNING: This is a DEVELOPMENT-MODE encryption implementation
 * Production deployment MUST use RescueCipher from Arcium SDK
 *
 * This module provides:
 * - ChaCha20-Poly1305 AEAD encryption
 * - HKDF-based key derivation per user
 * - Compatible with Rust implementation in arcium-service
 *
 * Security model:
 * - Master key from environment
 * - Per-user keys derived via HKDF(master_key, user_pubkey)
 * - Random nonces for each encryption operation
 * - Authenticated encryption prevents tampering
 *
 * Data format: [nonce (12 bytes)] + [ciphertext] + [auth tag (16 bytes)]
 */

const NONCE_LENGTH = 12; // ChaCha20-Poly1305 standard nonce length
const TAG_LENGTH = 16; // Authentication tag length
const KEY_LENGTH = 32; // 256-bit key

/**
 * Encryption helper for MPC computations
 *
 * Matches the Rust implementation in services/arcium-service/src/mpc/encryption.rs
 */
export class EncryptionHelper {
  private masterKey: Uint8Array;

  /**
   * Create encryption helper with master key
   *
   * @param masterKey - 32-byte (64 hex chars) master encryption key
   */
  constructor(masterKey: string) {
    // Convert hex string to bytes
    if (masterKey.length !== 64) {
      throw new Error('Master key must be 64 hex characters (32 bytes)');
    }
    this.masterKey = hexToBytes(masterKey);
  }

  /**
   * Derive a user-specific encryption key using HKDF
   *
   * Derivation: HKDF-SHA256(master_key, salt=user_pubkey, info="ninjapay-dev-v1")
   *
   * This MUST match the Rust implementation exactly
   */
  private deriveUserKey(userPubkey: string): Uint8Array {
    const salt = new TextEncoder().encode(userPubkey);
    const info = new TextEncoder().encode('ninjapay-dev-v1');

    // HKDF-SHA256 with salt and info
    const derivedKey = hkdf(sha256, this.masterKey, salt, info, KEY_LENGTH);

    return derivedKey;
  }

  /**
   * Encrypt a u64 value (for simulator)
   *
   * Converts number to 8-byte little-endian, then encrypts
   * Returns: [nonce (12)] + [ciphertext (8)] + [tag (16)] = 36 bytes total
   *
   * @param value - Number to encrypt (must fit in u64)
   * @param userPubkey - User's public key for key derivation
   * @returns Base64-encoded encrypted data
   */
  encryptU64(value: number | bigint, userPubkey: string): string {
    // Convert to 8-byte little-endian
    const plaintext = new Uint8Array(8);
    const view = new DataView(plaintext.buffer);
    view.setBigUint64(0, BigInt(value), true); // true = little-endian

    const encrypted = this.encryptBytes(plaintext, userPubkey);
    return bytesToBase64(encrypted);
  }

  /**
   * Decrypt to u64 value (for simulator)
   *
   * Decrypts and converts 8-byte little-endian to number
   *
   * @param encrypted - Base64-encoded encrypted data
   * @param userPubkey - User's public key for key derivation
   * @returns Decrypted number
   */
  decryptToU64(encrypted: string, userPubkey: string): bigint {
    const encryptedBytes = base64ToBytes(encrypted);
    const plaintext = this.decryptBytes(encryptedBytes, userPubkey);

    if (plaintext.length !== 8) {
      throw new Error(
        `Invalid u64 decryption: expected 8 bytes, got ${plaintext.length}`
      );
    }

    const view = new DataView(plaintext.buffer);
    return view.getBigUint64(0, true); // true = little-endian
  }

  /**
   * Encrypt arbitrary bytes
   *
   * Format: [nonce (12 bytes)] + [ciphertext] + [tag (16 bytes)]
   *
   * @param data - Data to encrypt
   * @param userPubkey - User's public key for key derivation
   * @returns Encrypted bytes (nonce + ciphertext + tag)
   */
  encryptBytes(data: Uint8Array, userPubkey: string): Uint8Array {
    // Derive user-specific key
    const userKey = this.deriveUserKey(userPubkey);

    // Generate random nonce
    const nonce = randomBytes(NONCE_LENGTH);

    // Create cipher instance
    const cipher = chacha20poly1305(userKey, nonce);

    // Encrypt (returns ciphertext + tag combined)
    const ciphertext = cipher.encrypt(data);

    // Combine: nonce + ciphertext
    const result = new Uint8Array(NONCE_LENGTH + ciphertext.length);
    result.set(nonce, 0);
    result.set(ciphertext, NONCE_LENGTH);

    return result;
  }

  /**
   * Decrypt arbitrary bytes
   *
   * Expects format: [nonce (12 bytes)] + [ciphertext] + [tag (16 bytes)]
   *
   * @param encrypted - Encrypted data
   * @param userPubkey - User's public key for key derivation
   * @returns Decrypted plaintext
   */
  decryptBytes(encrypted: Uint8Array, userPubkey: string): Uint8Array {
    // Validate minimum length (nonce + tag)
    const minLength = NONCE_LENGTH + TAG_LENGTH;
    if (encrypted.length < minLength) {
      throw new Error(
        `Invalid encrypted data: too short (got ${encrypted.length} bytes, need at least ${minLength})`
      );
    }

    // Extract nonce and ciphertext
    const nonce = encrypted.slice(0, NONCE_LENGTH);
    const ciphertext = encrypted.slice(NONCE_LENGTH);

    // Derive user-specific key
    const userKey = this.deriveUserKey(userPubkey);

    // Create cipher instance
    const cipher = chacha20poly1305(userKey, nonce);

    // Decrypt (will throw if authentication fails)
    try {
      const plaintext = cipher.decrypt(ciphertext);
      return plaintext;
    } catch (error) {
      throw new Error(`Decryption failed: ${error instanceof Error ? error.message : 'unknown error'}`);
    }
  }

  /**
   * Validate encrypted input format
   *
   * Checks:
   * 1. Not empty
   * 2. Minimum length (nonce + tag = 28 bytes)
   * 3. Proper structure
   *
   * @param encryptedData - Data to validate
   * @returns true if valid format
   */
  validateEncryptedInput(encryptedData: Uint8Array): boolean {
    if (encryptedData.length === 0) {
      console.warn('Validation failed: empty input');
      return false;
    }

    // Minimum: nonce (12) + minimal plaintext (0) + tag (16) = 28 bytes
    const minLength = NONCE_LENGTH + TAG_LENGTH;
    if (encryptedData.length < minLength) {
      console.warn(
        `Validation failed: too short (${encryptedData.length} bytes, need at least ${minLength})`
      );
      return false;
    }

    // For u64 encrypted values, expect exactly 36 bytes
    // nonce (12) + plaintext (8) + tag (16) = 36
    if (encryptedData.length === 36) {
      return true;
    }

    // Other lengths are valid as long as >= 28
    return true;
  }

  /**
   * Prepare encrypted inputs for batching
   *
   * Format: [count (4 bytes)] + [len1 (4)] + [data1] + [len2 (4)] + [data2] + ...
   *
   * @param encryptedValues - Array of encrypted values
   * @returns Batched encrypted data
   */
  prepareBatchInputs(encryptedValues: Uint8Array[]): Uint8Array {
    // Validate all inputs first
    for (let i = 0; i < encryptedValues.length; i++) {
      if (!this.validateEncryptedInput(encryptedValues[i])) {
        throw new Error(`Invalid encrypted input at index ${i}`);
      }
    }

    // Calculate total size
    const count = encryptedValues.length;
    let totalSize = 4; // count
    for (const value of encryptedValues) {
      totalSize += 4 + value.length; // len + data per item
    }

    const result = new Uint8Array(totalSize);
    const view = new DataView(result.buffer);
    let offset = 0;

    // Write count
    view.setUint32(offset, count, true); // little-endian
    offset += 4;

    // Write each value with length prefix
    for (const value of encryptedValues) {
      view.setUint32(offset, value.length, true); // little-endian
      offset += 4;
      result.set(value, offset);
      offset += value.length;
    }

    return result;
  }

  /**
   * Extract individual results from batch computation
   *
   * Expects format: [count (4 bytes)] + [len1 (4)] + [data1] + [len2 (4)] + [data2] + ...
   *
   * @param batchResult - Batched result data
   * @param expectedCount - Expected number of results
   * @returns Array of individual encrypted results
   */
  extractBatchResults(
    batchResult: Uint8Array,
    expectedCount: number
  ): Uint8Array[] {
    if (batchResult.length < 4) {
      throw new Error('Batch result too short: missing count');
    }

    const view = new DataView(batchResult.buffer);
    let offset = 0;

    // Read count
    const count = view.getUint32(offset, true); // little-endian
    offset += 4;

    if (count !== expectedCount) {
      throw new Error(
        `Batch count mismatch: expected ${expectedCount}, got ${count}`
      );
    }

    const results: Uint8Array[] = [];

    // Read each value
    for (let i = 0; i < count; i++) {
      if (offset + 4 > batchResult.length) {
        throw new Error(`Batch result truncated at item ${i}`);
      }

      // Read length
      const len = view.getUint32(offset, true); // little-endian
      offset += 4;

      // Read data
      if (offset + len > batchResult.length) {
        throw new Error(`Batch result truncated: item ${i} data incomplete`);
      }

      const data = batchResult.slice(offset, offset + len);
      results.push(data);
      offset += len;
    }

    return results;
  }
}

/**
 * Utility functions for HTTP API integration
 */
export class EncryptionAPIUtils {
  private encryption: EncryptionHelper;

  constructor(masterKey: string) {
    this.encryption = new EncryptionHelper(masterKey);
  }

  /**
   * Encrypt value for API request
   *
   * @param value - Number to encrypt
   * @param userPubkey - User's public key
   * @returns Base64-encoded encrypted value (ready for JSON/HTTP)
   */
  encryptForAPI(value: number | bigint, userPubkey: string): string {
    return this.encryption.encryptU64(value, userPubkey);
  }

  /**
   * Decrypt value from API response
   *
   * @param encrypted - Base64-encoded encrypted value
   * @param userPubkey - User's public key
   * @returns Decrypted number
   */
  decryptFromAPI(encrypted: string, userPubkey: string): bigint {
    return this.encryption.decryptToU64(encrypted, userPubkey);
  }

  /**
   * Prepare multiple encrypted inputs for API request
   *
   * @param values - Values to encrypt
   * @param userPubkey - User's public key
   * @returns Array of base64-encoded encrypted values
   */
  encryptBatchForAPI(
    values: (number | bigint)[],
    userPubkey: string
  ): string[] {
    return values.map((v) => this.encryption.encryptU64(v, userPubkey));
  }
}

/**
 * Utility: Convert hex string to bytes
 */
function hexToBytes(hex: string): Uint8Array {
  if (hex.length % 2 !== 0) {
    throw new Error('Hex string must have even length');
  }
  const bytes = new Uint8Array(hex.length / 2);
  for (let i = 0; i < bytes.length; i++) {
    bytes[i] = parseInt(hex.substr(i * 2, 2), 16);
  }
  return bytes;
}

/**
 * Utility: Convert bytes to base64
 */
function bytesToBase64(bytes: Uint8Array): string {
  if (typeof Buffer !== 'undefined') {
    // Node.js
    return Buffer.from(bytes).toString('base64');
  } else {
    // Browser
    let binary = '';
    for (let i = 0; i < bytes.length; i++) {
      binary += String.fromCharCode(bytes[i]);
    }
    return btoa(binary);
  }
}

/**
 * Utility: Convert base64 to bytes
 */
function base64ToBytes(base64: string): Uint8Array {
  if (typeof Buffer !== 'undefined') {
    // Node.js
    return new Uint8Array(Buffer.from(base64, 'base64'));
  } else {
    // Browser
    const binary = atob(base64);
    const bytes = new Uint8Array(binary.length);
    for (let i = 0; i < binary.length; i++) {
      bytes[i] = binary.charCodeAt(i);
    }
    return bytes;
  }
}

/**
 * Example usage for merchant dashboard:
 *
 * ```typescript
 * // Initialize encryption helper
 * const masterKey = process.env.ENCRYPTION_MASTER_KEY!;
 * const apiUtils = new EncryptionAPIUtils(masterKey);
 *
 * // Encrypt payment amount
 * const userPubkey = 'user_wallet_address';
 * const amount = 1000; // $10.00 in cents
 * const encryptedAmount = apiUtils.encryptForAPI(amount, userPubkey);
 *
 * // Send to Arcium Service API
 * const response = await fetch('http://localhost:8001/api/computation/invoke', {
 *   method: 'POST',
 *   headers: { 'Content-Type': 'application/json' },
 *   body: JSON.stringify({
 *     computation_type: 'encrypted_transfer',
 *     encrypted_inputs: [encryptedAmount],
 *     user_pubkey: userPubkey,
 *   }),
 * });
 *
 * // Get result
 * const { computation_id } = await response.json();
 *
 * // Poll for result
 * const resultResponse = await fetch(
 *   `http://localhost:8001/api/computation/status?computation_id=${computation_id}`
 * );
 * const { result } = await resultResponse.json();
 *
 * // Decrypt result
 * const decryptedResult = apiUtils.decryptFromAPI(result, userPubkey);
 * console.log('New balance:', decryptedResult);
 * ```
 */
